Serialize CryptoKey using CoreIPC

nmahendru opened the pull request:
#### f4eb2677c3861070dd00753bc2503a606b96d92b
<pre>
Serialize CryptoKey using CoreIPC
<a href="https://bugs.webkit.org/show_bug.cgi?id=272064">https://bugs.webkit.org/show_bug.cgi?id=272064</a>
<a href="https://rdar.apple.com/128551948">rdar://128551948</a>

Reviewed by Alex Christensen and David Kilzer.

Over the wire bytes will now be serialized using CoreIPC ::decode
and thus the crash in PropertyListDeserialization is avoided.
Note that serialization is still property list and that maintains backward
compatibility with serialized CryptoKeys on disks of existing users.

Testing: No new tests are added as this functionality is already being exercised by
Indexeddb tests/service worker tests.

* Source/WebCore/Headers.cmake:
* Source/WebCore/crypto/WrappedCryptoKey.h:
* Source/WebCore/crypto/SerializedCryptoKeyWrap.h:
* Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
(WebCore::wrapSerializedCryptoKey):
(WebCore::createArrayFromData):
(WebCore::readSerializedCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::unwrapSerializedCryptoKey): Deleted.
* Source/WebCore/crypto/gcrypt/SerializedCryptoKeyWrapGCrypt.cpp:
(WebCore::readSerializedCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::unwrapSerializedCryptoKey): Deleted.
* Source/WebCore/crypto/openssl/SerializedCryptoKeyWrapOpenSSL.cpp:
(WebCore::readSerializedCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::unwrapSerializedCryptoKey): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebCryptoClient.cpp:
(WebKit::WebCryptoClient::unwrapCryptoKey const):
* Source/WebKitLegacy/WebCoreSupport/WebCryptoClient.mm:
(WebCryptoClient::unwrapCryptoKey const):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/SerializedCryptoKeyWrap.mm:
(TestWebKitAPI::TEST_F):

Originally-landed-as: 272448.867@safari-7618-branch (b4edd8f7193f). <a href="https://rdar.apple.com/128551948">rdar://128551948</a>

test

Canonical link: <a href="https://commits.webkit.org/279933@main">https://commits.webkit.org/279933@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3932e839eb4bb5d0c5fed2a040f6f13ea76dfeba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54884 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34329 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7470 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58162 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5615 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57184 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5639 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44447 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3802 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56979 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32433 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47551 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25574 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29223 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; 4 new passes 3 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4929 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3756 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51070 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5111 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59752 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30140 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5259 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51870 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31278 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47629 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51287 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12082 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32292 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31066 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/crypto/CryptoKey.h:
```diff
@@ -92,8 +92,15 @@ inline auto CryptoKey::type() const -> Type
+
 } // namespace WebCore
 
+
```
Super nit:  Don't need extra space here.  (Not required to fix to land the PR.)

ddkilzer commented on Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
```diff
@@ -61,7 +62,11 @@
+constexpr size_t kekSizeInBytes = 16;
+constexpr size_t expectedTagLengthAES = 16;
+// https://datatracker.ietf.org/doc/html/rfc3394#section-2.2.1
+constexpr size_t wrappedKekSize = kekSizeInBytes + 8;
```
Nit:  I can't tell if this comment is for the previous line or the next line.  Might be best to put it on the same line as what it's documenting.  (Not required to land the PR.)

ddkilzer commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -8236,3 +8236,11 @@ enum class WebCore::CrossOriginMode : bool
+    Vector<uint8_t> encryptedKey;
+    std::array<uint8_t, 16> tag;
+};
+
```
Super nit:  Don't need blank line at the end of the file.  (Not required to land PR.)

Super nit # 2:  I thought this made sense to put closer to where it landed on the branch rather than the very end.  (Again, not needed to land PR.)

ddkilzer commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -93,6 +93,7 @@ struct PluginInfo;
 struct WebProcessCreationParameters;
 class SecurityOriginData;
+struct WrappedCryptoKey;
 enum class PermissionName : uint8_t;
```
Nit:  Should have a space after the last `struct` declaration.

ddkilzer commented on Source/WebKitLegacy/WebCoreSupport/WebCryptoClient.mm:
```diff
@@ -50,21 +51,25 @@
+        auto key = WebCore::unwrapCryptoKey(masterKey, *wrappedKey);
+        if (!key)
             return std::nullopt;
         return key;
```
It's a bit more idiomatic (closer to WebKit style) to reverse these two lines (and since getting a key is the more likely code path):

```suggestion
        if (auto key = WebCore::unwrapCryptoKey(masterKey, *wrappedKey))
            return key;
        return std::nullopt;
```

I think this also works, but please check first:

```suggestion
        return WebCore::unwrapCryptoKey(masterKey, *wrappedKey) ?: std::nullopt;
```

Or if `WebCore::unwrapCryptoKey()` already returns `std::nullopt` on error, just:

```suggestion
        return WebCore::unwrapCryptoKey(masterKey, *wrappedKey);
```

nmahendru replied:
Excellent suggestion!!!

ddkilzer commented on Source/WebKitLegacy/WebCoreSupport/WebCryptoClient.mm:
```diff
@@ -50,21 +51,25 @@
+    auto key = WebCore::unwrapCryptoKey(WTFMove(*masterKey), *wrappedKey);
+    if (!key)
         return std::nullopt;
     return key;
```
Ditto to previous comment.

ddkilzer reviewed (approved):
r=me.  The code is fine as-is, but I'd like to see the logic for the last two comments simplified.

nmahendru committed:
Serialize CryptoKey using CoreIPC
https://bugs.webkit.org/show_bug.cgi?id=272064
rdar://128551948

Reviewed by Alex Christensen and David Kilzer.

Over the wire bytes will now be serialized using CoreIPC ::decode
and thus the crash in PropertyListDeserialization is avoided.
Note that serialization is still property list and that maintains backward
compatibility with serialized CryptoKeys on disks of existing users.

Testing: No new tests are added as this functionality is already being exercised by
Indexeddb tests/service worker tests.

* Source/WebCore/Headers.cmake:
* Source/WebCore/crypto/WrappedCryptoKey.h:
* Source/WebCore/crypto/SerializedCryptoKeyWrap.h:
* Source/WebCore/crypto/cocoa/SerializedCryptoKeyWrapMac.mm:
(WebCore::wrapSerializedCryptoKey):
(WebCore::createArrayFromData):
(WebCore::readSerializedCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::unwrapSerializedCryptoKey): Deleted.
* Source/WebCore/crypto/gcrypt/SerializedCryptoKeyWrapGCrypt.cpp:
(WebCore::readSerializedCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::unwrapSerializedCryptoKey): Deleted.
* Source/WebCore/crypto/openssl/SerializedCryptoKeyWrapOpenSSL.cpp:
(WebCore::readSerializedCryptoKey):
(WebCore::unwrapCryptoKey):
(WebCore::unwrapSerializedCryptoKey): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebCryptoClient.cpp:
(WebKit::WebCryptoClient::unwrapCryptoKey const):
* Source/WebKitLegacy/WebCoreSupport/WebCryptoClient.mm:
(WebCryptoClient::unwrapCryptoKey const):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/SerializedCryptoKeyWrap.mm:
(TestWebKitAPI::TEST_F):

Originally-landed-as: 272448.867@safari-7618-branch (b4edd8f7193f). rdar://128551948

test

Canonical link: https://commits.webkit.org/279933@main
